content/post/elm-at-rakuten/index.md

```diff
@@ -95,82 +103,109 @@ In no particular order.
-
-[^type-guards]: [Type guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html) are TypeScript expressions that perform a runtime check to discriminate between *custom types*. For example:
-{{< highlight typescript >}}
-function isFish(pet: Fish | Bird): pet is Fish {
-    return (pet as Fish).swim !== undefined;
-}
-{{< /highlight >}}
-After that is possible to write code like:
-{{< highlight javascript >}}
-if (isFish(pet)) {
+TypeScript, which adds optional static typing to JavaScript and is probably one of the best things to happen to JavaScript, can partially mitigate the issues with the JavaScript dynamic type system. But being a superset of JavaScript it needs to compromise on elegance and simplicity. It also has several "blind spots".[^blind-spots] For example, type declarations are optional (_any_ as escape hatch), inference doesn’t cover all the code, it requires _type guards_,[^type-guards] JSON data is not type-checked, and not all JavaScript libraries have type annotations.
+
+[^blind-spots]: The article [TypeScript's Blind Spots](https://incrementalelm.com/tips/typescript-blind-spots/) by Dillon Kearns illustrates several weak points of TypeScript.
+[^type-guards]:
+    [Type guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html) are TypeScript expressions that perform a runtime check to discriminate between _custom types_. For example:
+    {{< highlight typescript >}}function isFish(pet: Fish | Bird): pet is Fish {
+    return (pet as Fish)
+    .swim !== undefined;
+    }{{< /highlight >}}
+    After that is possible to write code like:{{< highlight javascript >}}if (isFish(pet)) {
     pet.swim();
-} else {
+    } else {
     pet.fly();
-}
-{{< /highlight >}}
-In Elm *type guards* are not needed, it is possible to just directly use a [case .. of](https://guide.elm-lang.org/types/pattern_matching.html) construct
-{{< highlight elm >}}
-case pet of
-    Fish fish ->
-        fish.swim
+    }{{< /highlight >}}In Elm _type guards_ are not needed, it is possible to just directly use a [case .. of](https://guide.elm-lang.org/types/pattern_matching.html) construct{{< highlight elm >}}case pet of
+    Fish fish -> fish.swim
+    Bird bird -> bird.fly{{< /highlight >}}The **case .. of** construct also guarantees that we are considering all possible custom types.
 
-    Bird bird ->
-        bird.fly
-{{< /highlight >}}
-The **case .. of** construct also guarantees that we are considering all possible custom types.
+[Index](#index) | [Top](#top)
 
-## 3. “Making impossible states impossible”
+---
+
+## 4. “Making impossible states impossible”
 
 ![Penrose Triangle](./images/penrose-triangle.jpg)
-*Penrose Triangle.*[^penrose-triangle]
+_Penrose Triangle.[^penrose-triangle]_
```

Comment:
The Penrose triangle illustration has no license information.

Reply:
This is original, made by me